### ISSUE TITLE
Type captured by wildcard pattern is partially unknown - Why?

### ISSUE THREAD

**Issue by carlos-souto** [microsoft/pylance-release#4058]
## Environment data

-   Language Server version: 2023.3.20
-   OS and version: Windows 11 x64
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.11.2 64-bit

## Code Snippet

```python
print('LIST OF OPERATIONS:')
print('0: ADD')
print('1: SUBTRACT')
print('2: MULTIPLY')
print('3: DIVIDE')

operation: str = ''
a = 0
b = 0
for i in range(3):

    if i == 0:
        user: str = input('STEP 1, CHOOSE OPERATION: ')
        operation = (
            'ADD'      if user == '0' else
            'SUBTRACT' if user == '1' else
            'MULTIPLY' if user == '2' else
            'DIVIDE'   if user == '3' else
            ''
        )

    elif i == 1:
        print('STEP 2, CHOOSE a AND b')
        try:
            a: int = int(input('a = '))
        except:
            print('invalid a, assuming 0')
        try:
            b: int = int(input('b = '))
        except:
            print('invalid b, assuming 0')

    else:
        print('STEP 3, RESULT:')
        match operation:                     # (variable) operation: Literal['', 'ADD', 'SUBTRACT', 'MULTIPLY', 'DIVIDE']
            case 'ADD':      print(a + b)
            case 'SUBTRACT': print(a - b)
            case 'MULTIPLY': print(a * b)
            case 'DIVIDE':   print(a / b)
            case _:                          # Type captured by wildcard pattern is partially unknown (why?)
                print('invalid operation')

```

In the match statement, the variable `operation` is correctly identified as `(variable) operation: Literal['', 'ADD', 'SUBTRACT', 'MULTIPLY', 'DIVIDE']`. So why is pylance reporting the following in the `case _:` case statement:

```
_: Unknown
Type captured by wildcard pattern is partially unknown
Type of "_" is "Unknown | Literal['']"
```

![image](https://user-images.githubusercontent.com/83190503/224451542-f71a90d3-77e6-4969-ac95-96ce3fdbd774.png)

![image](https://user-images.githubusercontent.com/83190503/224451451-f7dafd60-a666-4223-8d31-8385b64fae48.png)

My expectation is that no error should be reported.

Please note that I am not implying that the code snippet shown above makes very much sense. The for loop is unnecessary; however, I found that it's the loop that creates this unwanted behavior in the `case _:` statement. The code snippet is just a minimal working example that showcases the issue that I have in a real-world piece of code.

## Logs

Attached.
[log.txt](https://github.com/microsoft/pylance-release/files/10946966/log.txt)

**Comment by erictraut** [comment]
Yes, this looks like a bug in the type narrowing logic for pattern matching. I'll investigate.

**Comment by carlos-souto** [comment]
I believe this issue appeared with the new update of Pylance (released on 2023-03-09). Before that, I didn't have this problem.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in a future release.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.22, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023322-15-March-2023-PreRelease)
